### ISSUE TITLE
generic constructors as arguments are unchecked

### ISSUE THREAD

**Issue by kwshi** [microsoft/pyright#7695]
**Describe the bug**
The type `type[C]` denotes the type of the _constructor_ of `C`; e.g., the literal expression `int` has type `type[int]`.

In particular, we can define functions using the signature `def f[C](constructor: type[C])` to mean "`f` is a generic function that takes any class constructor function as an argument". Inside `f`, we can then invoke the constructor with various arguments; however there seems to be no type-checking as to whether those arguments are actually valid/accepted by the constructor.

**Code or Screenshots**

```python
# pyright: strict

class X:
    pass

def f[C](constructor: type[C]):
    constructor(1, 2)

f(X)
```

```
Traceback (most recent call last):
  File "/home/kshi/stuff.py", line 12, in <module>
    construct(X)
  File "/home/kshi/stuff.py", line 9, in construct
    cls(1, 2)
TypeError: X() takes no arguments
```

**VS Code extension or command-line**
CLI/nvim LSP; version 1.1.354.

**Comment by erictraut** [comment]
Thanks for the bug report. This behavior was previously unspecified in the Python typing spec, but a few days ago, a new chapter (that I authored) was accepted into the typing spec, and it clarifies the behavior [in this case](https://typing.readthedocs.io/en/latest/spec/constructors.html#constructor-calls-for-type-t). You are correct that pyright doesn't currently handle this case correctly. I'll work on a fix.

**Comment by erictraut** [comment]
Pyright previously exempted `type[T]` (where `T` has no explicit upper bound) from enforcing the `__new__` and `__init__` signatures of `object`. The typing spec now clarifies that it should enforce the `object` signature in this case. This is consistent with mypy's current behavior, but this change is going to cause some temporary pain for existing pyright users. 

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.359, which I just published.
